app/lib/task/backend.dart

```diff
@@ -1048,16 +1055,48 @@ class TaskBackend {
         if (bestVersion != null) {
           return bestVersion.toString();
         }
-        // TODO: remove this fallback after all runtime version has the `finished` field populated
-        if (rt.compareTo('2023.08.29') < 0) {
-          final latestNonPending = state.versions?.entries
-              .where((e) => e.value.status != PackageVersionStatus.pending)
-              .map((e) => Version.parse(e.key))
-              .latestVersion;
-          if (latestNonPending != null) {
-            return latestNonPending.toString();
-          }
+      }
+      return '';
+    });
+    return (cachedValue == null || cachedValue.isEmpty) ? null : cachedValue;
+  }
+
+  /// Returns the closest version of the [package] which has a finished analysis.
+  ///
+  /// If [version] or new exists with finished analysis, it will be preferred, otherwise
+  /// older versions may be considered too.
+  ///
+  /// Returns `null` if no such version exists.
+  Future<String?> closestFinishedVersion(String package, String version) async {
+    final cachedValue =
+        await cache.closestFinishedVersion(package, version).get(() async {
+      final semanticVersion = Version.parse(version);
+      for (final rt in acceptedRuntimeVersions) {
+        final key = PackageState.createKey(_db, rt, package);
+        final state = await dbService.lookupOrNull<PackageState>(key);
+        // skip states where the entry was created, but no analysis has not finished yet
+        if (state == null || state.hasNeverFinished) {
+          continue;
+        }
+        final candidates = state.versions?.entries
+            .where((e) => e.value.finished)
+            .map((e) => Version.parse(e.key))
+            .toList();
+        if (candidates == null || candidates.isEmpty) {
+          continue;
+        }
+        if (candidates.contains(semanticVersion)) {
+          return version;
+        }
+        final newerCandidates =
+            candidates.where((e) => isNewer(semanticVersion, e)).toList();
+        if (newerCandidates.isNotEmpty) {
+          // return the earliest finished that is newer than [version].
+          return newerCandidates
+              .reduce((a, b) => isNewer(a, b) ? a : b)
+              .toString();
         }
+        return candidates.latestVersion!.toString();
       }
       return '';
```

Comment:
Just to clarify my understanding: Given that you return empty string here representing "computed, but found no value" can `cachedValue` ever be `null` below?

Reply:
Correct. `null` values won't be cached, triggering recompute.